lib/Driver/Driver.cpp

```diff
@@ -1393,27 +1389,33 @@ void Driver::buildOutputInfo(const ToolChain &TC, const DerivedArgList &Args,
 
 OutputInfo::Mode
 Driver::computeCompilerMode(const DerivedArgList &Args,
-                            const InputFileList &Inputs) const {
+                            const InputFileList &Inputs,
+                            bool &BatchModeOut) const {
 
   if (driverKind == Driver::DriverKind::Interactive)
     return Inputs.empty() ? OutputInfo::Mode::REPL
                           : OutputInfo::Mode::Immediate;
 
-  const Arg *ArgRequiringWMO = Args.getLastArg(
+  const Arg *ArgRequiringSingleCompile = Args.getLastArg(
       options::OPT_whole_module_optimization, options::OPT_index_file);
 
-  if (!ArgRequiringWMO)
-    return OutputInfo::Mode::StandardCompile;
-
-  // Test for -enable-batch-mode, rather than the BatchMode flag that is
-  // passed into the caller because the diagnostic is intended to warn against
-  // overriding *explicit* batch mode. No warning should be given if in batch
-  // mode by default.
-  if (Args.hasArg(options::OPT_enable_batch_mode))
-    Diags.diagnose(SourceLoc(), diag::warn_ignoring_batch_mode,
-                   ArgRequiringWMO->getOption().getPrefixedName());
+  BatchModeOut = Args.hasFlag(options::OPT_enable_batch_mode,
+                              options::OPT_disable_batch_mode,
+                              false);
+
+  if (ArgRequiringSingleCompile) {
+    // Override batch mode if given -wmo or -index-file.
+    if (BatchModeOut) {
+      BatchModeOut = false;
+      // Emit a warning about such overriding (FIXME: we might conditionalize
+      // this based on the user or xcode passing -disable-batch-mode).
+      Diags.diagnose(SourceLoc(), diag::warn_ignoring_batch_mode,
+                     ArgRequiringSingleCompile->getOption().getPrefixedName());
+    }
+    return OutputInfo::Mode::SingleCompile;
+  }
 
-  return OutputInfo::Mode::SingleCompile;
+  return OutputInfo::Mode::StandardCompile;
```

Comment:
If you like, you could clarify the code a smidge by changing 1406 to if (!ArgRequiringSingleCompile) and then have an immediate return StandardCompile after that if.

Reply:
That's what it used to do; I inverted the logic while I was trying to figure out the best way to factor this, and then realized I thought it read better that way so left it as-is. Do you feel strongly about my inverting it again?

Reply:
Doesn't matter. Just want to land this. I'll change it back.

Reply:
OK, tx. Would still like to understand what you see there that I don't. I believe it's real.